README
NAME:
     Geo::Heatmap
     
     overlay a density map (aka heatmap) over googlemaps

     see example here

     http://www.trust-box.at//dev/gm/GoogleMapsHeatmap/www/GoogleMapsHeatmap.html

GET CODE
    https://github.com/MarkHofstetter/GoogleMapsHeatmap

SYNOPSIS
    Create a Heatmap layer for GoogleMaps

    my $ghm = Geo::Heatmap->new();
    $ghm->palette('palette.store');
    $ghm->zoom_scale( {
      1 => 298983,
      2 => 177127,
      3 => 104949,
      4 => 90185,
      5 => 70338,
      6 => 37742,
      7 => 28157,
      8 => 12541,
      9 => 3662,
      10 => 1275,
      11 => 417,
      12 => 130,
      13 => 41,
      14 => 18,
      15 => 10,
      16 => 6,
      17 => 2,
      18 => 0,
    } );
  
    $ghm->cache($cache);
    $ghm->return_points( \&get_points );
    my $image = $ghm->tile($tile);


    You need a color palette (one is included) to encode values to colors, in Storable Format as an arrayref of arrayrefs eg
      [50] = [34, 45, 56]
    which means that a normalized value of 50 would lead to an RGB color of 34% red , 45% blue, 56% green

    zoom_scale
      the maximum number of points for a given google zoom scale, you would be able to extract to values from the denisity log
      or derive them from your data in some cunning way

    cache 
      you need some caching for the tiles otherwise the map would be quite slow. Use a CHI object with the cache you like

    return_points
      is a function reference which expects a single hashref as a parameter which defines two LAT/LONG points to get all
      data points within this box
      $r->{LATN}, $r->{LNGW}), $r->{LATS}, $r->{LNGE}
      the function has to return an arrayref of arrayrefs of the points within the box

    tile 
      returns the rendered image




Changes
Revision history for Geo::Heatmap

0.16 2013-12-05
        - Moved from Image::Magick to Imager
        - migrated HTML example to Google Map API V3


0.17  2013-12-07 06:04:59 CET
        - Example to Create Zoom Scale

{{$NEXT}}

0.18  2013-12-09 10:01:46 CET
        - Working on documentation
        - Setting things in order


lib/GoogleHeatmap.pm
package GoogleHeatmap;
use Moose;

# generate a density map (aka heatmap) overlay layer for Google Maps

# debug - if set to a value larger than 0 the package emits various debugging information
has 'debug'         => (isa => 'Str', is => 'rw');
has 'cache'         => (isa => 'Object', is => 'rw');
has 'logfile'       => (isa => 'Str', is => 'rw');
has 'return_points' => (isa => 'CodeRef', is => 'rw'); 
has 'zoom_scale'    => (isa => 'HashRef', is => 'rw'); 
has 'palette'       => (isa => 'Str', is => 'rw');

__PACKAGE__->meta->make_immutable;

use USNaviguide_Google_Tiles;
use Image::Magick;
use Storable;

sub tile {
  my ($self, $tile, $debug) = @_;
  $debug |= 0;

  my ($x, $y, $z) = split(/\s+/, $tile);

  my $e;
  my $image = Image::Magick->new(magick=>'png');
  my %ubblob;
  my $k = 0;
  my $stitch;
  my $wi;
  my $line;

  my $mca = sprintf("blur_%s_%s_%s", $x, $y, $z);
  my $ubblob = $self->cache->get($mca);
  return $ubblob if defined $ubblob;

  for (my $i = -1; $i <= 1; $i++) {
    my $li = Image::Magick->new();
    for (my $j = -1; $j <= 1; $j++) {
      $ubblob{$i}{$j} = $self->calc_hm_tile([$x+$j, $y+$i, $z], $debug);
      $li->BlobToImage($ubblob{$i}{$j});
    }
    $line = $li->Append(stack => 'false');
    push @$image, ($line);
  }
 
  $wi = $image->Append(stack => 'true');
  $wi->GaussianBlur(geometry=>'768x768', radius=>"6", sigma=>"4");
  $wi->Crop(geometry=>'255x255+256+256');
  if ($debug > 1) {
    print "Debugging stitch\n";
    $wi->Write($mca.".png");
  }
  
  $ubblob =  $wi->ImageToBlob();
  $self->cache->set($mca, $ubblob);
  return $ubblob;
}

sub calc_hm_tile {
  my ($self, $coord, $debug) = @_;

  my ($x, $y, $z) = @$coord;

  my $mca = sprintf("raw_%s_%s_%s", $x, $y, $z);
  my $ubblob = $self->cache->get($mca);
  return $ubblob if defined $ubblob;  
  my $zoom_scale = $self->zoom_scale;

  my $value = &Google_Tile_Factors($z, 0) ;
  my %r = Google_Tile_Calc($value, $y, $x);

  my $image = Image::Magick->new(magick=>'png');
  $image->Set(size=>'256x256');
  $image->ReadImage('xc:white');
  my $rp = $self->return_points();
  my $ps = &$rp(\%r);;
  my $palette = Storable::retrieve($self->palette);
  $palette->[-1] = [100, 100, 100];
  my @density;
  my $bin = 8;
  foreach my $p (@$ps) {
    my @d = Google_Coord_to_Pix($value, $p->[0], $p->[1]);
    my $ix = $d[1] - $r{PXW};
    my $iy = $d[0] - $r{PYN};
    $density[int($ix/$bin)][int($iy/$bin)] ++;
    printf "%s %s %s %s\n", $ix, $iy, $ix % $bin, $iy % $bin if $debug >5;
  }
  
  my $maxval = ($bin)**2;
  # $zoom_scale->{$z} |= 0;
  my $defscale = $zoom_scale->{$z} > 0 ? $zoom_scale->{$z} : $maxval;
  $defscale *= 1.1;
  my $scale = 500/log($defscale);
  my $max = 256/$bin - 1;
  my $dmax = 0;
  
  for (my $i = 0; $i <= $max; $i++) {
    for (my $j = 0; $j <= $max; $j++) {
      # $image->Draw(fill=>rgb("$density->[$i][$j], 0, 0") , primitive=>'rectangle', points=>"$i,$j $bin,$bin");
      my $xc  = $i*$bin;
      my $yc  = $j*$bin;
      my $xlc = $xc+$bin;
      my $ylc = $yc+$bin;
      my $d = $density[$i][$j] ? $density[$i][$j] : 1;
      $dmax = $d > $dmax ? $d : $dmax;
      my $color_index = int(500-log($d)*$scale);
      $color_index = 5 if $color_index < 1;
      $color_index = -1 if $d < 2;
      my $color = $palette->[$color_index];
      my $rgb = sprintf "%s%%, %s%%, %s%%", $color->[0], $color->[1], $color->[2];
      $image->Draw(fill=>"rgb($rgb)" , primitive=>'rectangle', points=>"$xc,$yc $xlc,$ylc");
    # $image->Draw(fill=>"rgb($rgb)" , primitive=>'circle', points=>"$xc,$yc $xlc,$ylc");
      printf "[%s][%s] %s %s\n", $i, $j, $d, $color_index if $debug > 0;
    }
  }
  
  if ($self->logfile) {
    open (LOG, ">>");
    printf LOG  "densitylog:  x y z pointcount: %s %s %s %s\n", $x, $y, $z, $dmax;
    close LOG;
  }
  
  ($ubblob) = $image->ImageToBlob();
  $self->cache->set($mca, $ubblob);
  return $ubblob;
}


sub create_tile {
  my ($tile) = @_;
  my ($x, $y, $z) = split(/\s+/, $tile);
  my $value      = &Google_Tile_Factors($z, 0) ;
  my %r = Google_Tile_Calc($value, $y, $x);

  my $image = Image::Magick->new(magick=>'png');
  $image->Set(size=>'256x256');
  $image->ReadImage('xc:white');
  my $ps = get_points(\%r);

  foreach my $p (@$ps) {
    my @d = Google_Coord_to_Pix($value, $p->[0], $p->[1]);
    my $ix = $d[1] - $r{PXW};
    my $iy = $d[0] - $r{PYN};
   ## $image->Set("pixel[$ix,$iy]"=>'red');
   ## $image->Set("pixel[$ix,$iy]"=>'green');
    $image->Set("pixel[$ix,$iy]"=>"rgb(0, 0, 89)");
  }

  my $nw = sprintf("%s %s", $r{LATN}, $r{LNGW});
  my $se = sprintf("%s %s", $r{LATS}, $r{LNGE});
  my ($blob) =  $image->ImageToBlob();
  return $blob;
}


1;


__END__

=pod

=head1 NAME

Geo::Maps::Heatmap::Google - generate a density map (aka heatmap) overlay layer for Google Maps

=head1 VERSION

version 0.5

=head1 REQUIRES

L<Storable>

L<CHI>

L<Image::Magick>

L<USNaviguide_Google_Tiles>

L<Moose>


=head1 METHODS

=head2 calc_hm_tile

 calc_hm_tile();

=head2 create_hm_tile

 create_hm_tile();

=head2 create_tile

 create_tile();


=head1 ATTRIBUTES

has 'debug'         => (isa => 'Str', is => 'rw');
has 'cache'         => (isa => 'Object', is => 'rw');
has 'logfile'       => (isa => 'Str', is => 'rw');
has 'return_points' => (isa => 'CodeRef', is => 'rw');
has 'zoom_scale'    => (isa => 'HashRef', is => 'rw');
has 'palette'       => (isa => 'Str', is => 'rw');

=head2 new

  my $ghm = CPAN::Uploader->new();

This method returns a new uploader.  You probably don't need to worry about
this method.

Valid arguments are the same as those to C<upload_file>.

=head2 read_config_file

  my $config = CPAN::Uploader->read_config_file( $filename );

This reads the config file and returns a hashref of its contents that can be
used as configuration for CPAN::Uploader.

If no filename is given, it looks for F<.pause> in the user's home directory
(from the env var C<HOME>, or the current directory if C<HOME> isn't set).

=head2 log

  $uploader->log($message);

This method logs the given string.  The default behavior is to print it to the
screen.  The message should not end in a newline, as one will be added as
needed.

=head2 log_debug

This method behaves like C<L</log>>, but only logs the message if the
CPAN::Uploader is in debug mode.

=head1 ORIGIN

This code is mostly derived from C<cpan-upload-http> by Brad Fitzpatrick, which
in turn was based on C<cpan-upload> by Neil Bowers.  I (I<rjbs>) didn't want to
have to use a C<system> call to run either of those, so I refactored the code
into this module.

=head1 AUTHOR

Mark Hofstetter <hofstettm@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Mark Hofstetter

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

=head1 REQUIRES

L<Storable> 

L<CHI> 

L<Image::Magick> 

L<USNaviguide_Google_Tiles> 

L<Moose> 


=head1 METHODS

=head2 calc_hm_tile

 calc_hm_tile();

=head2 create_hm_tile

 create_hm_tile();

=head2 create_tile

 create_tile();


=cut



script/max_points_per_tile_zoom.pl
use strict;
use warnings;
use Geo::Heatmap::USNaviguide_Google_Tiles;
use Data::Dumper;
use DBI;
use YAML;

my $dbh = DBI->connect("dbi:Pg:dbname=gisdb", 'gisdb', 'gisdb', {AutoCommit => 0});

## where are most of my points:
# we use ST_GeoHash with a length of 5 (about city size)
# to group the points
# as there may be points which ST_GeoHash can not handle we check the borders
# where most of the points are ie points with the same hash
# select the Box (ST_Extent) covering all this points
# to be more sure one could use more rows 


my $sth = $dbh->prepare(qq(
  with geohash as (
    select ST_GeoHash(geom::geometry, 5) st_geohash,
           geom
       from geodata where
          not(St_X(geom) < -180 or St_Y(geom)< -90 or St_X(geom) > 180 or St_Y(geom) > 90)
    )
  SELECT ST_Extent(geom) as extent FROM geohash 
    where st_geohash in
        (select st_geohash from 
           (select st_geohash, count(*) c 
              from geohash 
             group by st_geohash 
     order by c desc limit 2) max_geohash)
  )
);

$sth->execute();
my @p;
my @r = $sth->fetchrow;
my ($lats, $lngw, $latn, $lnge) = ($r[0] =~/BOX\((.+?) (.+?),(.+?) (.+?)\)/);
$sth->finish;

print STDERR "$lats, $lngw, $latn, $lnge\n";

## 48.164089, 16.3476612, 48.2080069, 16.391593
## my $lats = 48.164089;
## my $latn = 48.2080069;
## my $lngw = 16.3476612;
## my $lnge = 16.391593;

my @gt;

my $max_per_level = {}; 
for (my $zoomlevel = 0; $zoomlevel <=18; $zoomlevel++) 
{
  my $max = 0;
  $max_per_level->{$zoomlevel} = 0;
  @gt = Google_Tiles($lats, $lngw, $latn, $lnge, $zoomlevel);
  
  foreach my $tile (@gt) {
    $max = max_points_per_tile($tile->{NAMEX}, $tile->{NAMEY}, $zoomlevel);
    $max_per_level->{$zoomlevel} = $max if $max > $max_per_level->{$zoomlevel}; 
  }
  printf STDERR "Tile found [%s] Zoom [%s] max points [%s]\n", scalar @gt, $zoomlevel, $max;
}
print Dump $max_per_level;

sub max_points_per_tile {
  my ($x, $y, $z) = @_;

  my $value = &Google_Tile_Factors($z, 0) ;

  my %r = Google_Tile_Calc($value, $y, $x);

  my @density;
  my $bin = 8;
  my $ps = get_points(\%r);
  my $max = 0;
  
  foreach my $p (@$ps) {
    my @d = Google_Coord_to_Pix($value, $p->[0], $p->[1]);
    my $ix = $d[1] - $r{PXW};
    my $iy = $d[0] - $r{PYN};
    $density[int($ix/$bin)][int($iy/$bin)] ++;
    $max = $density[int($ix/$bin)][int($iy/$bin)] if $density[int($ix/$bin)][int($iy/$bin)] > $max;
  }
  return $max;
}

sub get_points {
  my $r = shift;
  my $sth = $dbh->prepare( qq(select ST_AsEWKT(geom) from geodata
                         where geom &&
              ST_SetSRID(ST_MakeBox2D(ST_Point($r->{LATN}, $r->{LNGW}),
                                      ST_Point($r->{LATS}, $r->{LNGE})
                        ),4326))
              );

  $sth->execute();
  my @p;
  while (my @r = $sth->fetchrow) {
    my ($x, $y) = ($r[0] =~/POINT\((.+?) (.+?)\)/);
    push (@p, [$x ,$y]);
  }
  $sth->finish;
  return \@p;
}



lib/Geo/Heatmap.pm
package Geo::Heatmap;
use Moose;
use Geo::Heatmap::USNaviguide_Google_Tiles;
use Imager;
use Storable;

has 'debug'         => (isa => 'Str', is => 'rw', default => 0);
has 'cache'         => (isa => 'Object', is => 'rw');
has 'logfile'       => (isa => 'Str', is => 'rw');
has 'return_points' => (isa => 'CodeRef', is => 'rw'); 
has 'zoom_scale'    => (isa => 'HashRef', is => 'rw'); 
has 'palette'       => (isa => 'Str', is => 'rw');
has 'scale'         => (isa => 'Num', is => 'rw', default => 1);
has 'blur'          => (isa => 'Int', is => 'rw', default => 4);  
has 'bin'           => (isa => 'Int', is => 'rw', default => 8); # should be a divisor of 256 
has 'density'       => (isa => 'HashRef', is => 'rw'); 

our $VERSION = '0.19';

__PACKAGE__->meta->make_immutable;

sub tile {
  my ($self, $tile, $debug) = @_;
  $debug |= 0;

  my ($x, $y, $z) = split(/\s+/, $tile);

  my $e;
  my %ubblob;
  my $k = 0;
  my $stitch;
  my $wi;
  my $line;
  my $ts = 256;

  my $mca = sprintf("blur_%s_%s_%s", $x, $y, $z);
  my $ubblob = $self->cache->get($mca);
  return $ubblob if defined $ubblob;

  my $image = Imager->new(xsize=>768, ysize=>768);
  for (my $i = -1; $i <= 1; $i++) {
    my $li = Imager->new();
    for (my $j = -1; $j <= 1; $j++) {
      $ubblob{$i}{$j} = $self->calc_hm_tile([$x+$j, $y+$i, $z]);
      $li->read(data => $ubblob{$i}{$j} ); 
  #    printf "%s %s\n", ($i+1)*$ts, ($j+1)*$ts;
      $image->paste(left=> ($j+1)*$ts, top=> ($i+1)*$ts, img=>$li);
    }
  }

  $image->filter(type=>"gaussian", stddev=>$self->blur);
  ##$image->filter(type=>"gradgen", 
  ##             xo=>[ 1, 1, 1 ], 
  ##             yo=>[ 1, 1, 1 ],
  ##             colors=>[ qw(red blue green) ]);
  my $cropped = $image->crop(left=>255, top=>255, width=>256, height=>256);
  
  if ($debug > 1) {
    print "Debugging stitch\n";
  }

  $cropped->write(data => \$ubblob, type => 'png');
  $self->cache->set($mca, $ubblob);
  return $ubblob;
}


sub calc_hm_tile {
  my ($self, $coord) = @_;

  my ($x, $y, $z) = @$coord;

  my $mca = sprintf("raw_%s_%s_%s", $x, $y, $z);
  my $ubblob = $self->cache->get($mca);
  return $ubblob if defined $ubblob;  
  my $zoom_scale = $self->zoom_scale;

  my $value = &Google_Tile_Factors($z, 0) ;
  my %r = Google_Tile_Calc($value, $y, $x);

  my $image = Imager->new(xsize=>256, ysize=>256);
  $image->box(filled => 1, color => 'white');
  my $rp = $self->return_points();
  my $ps = &$rp(\%r);;
  my $palette = Storable::retrieve($self->palette);
  $palette->[-1] = [100, 100, 100];
  my @density;
  my $bin = $self->bin;
  foreach my $p (@$ps) {
    my @d = Google_Coord_to_Pix($value, $p->[0], $p->[1]);
    my $ix = $d[1] - $r{PXW};
    my $iy = $d[0] - $r{PYN};
    $density[int($ix/$bin)][int($iy/$bin)] ++;
    printf "%s %s %s %s\n", $ix, $iy, $ix % $bin, $iy % $bin if $self->debug >5;
  }
  
  my $maxval = ($bin)**2;
  # $zoom_scale->{$z} |= 0;
  my $defscale = $zoom_scale->{$z} > 0 ? $zoom_scale->{$z} : $maxval;
  $defscale *= 1.1 * $self->scale;
  my $scale = 500/log($defscale);
  my $max = 256/$bin - 1;
  my $dmax = 0;
  
  for (my $i = 0; $i <= $max; $i++) {
    for (my $j = 0; $j <= $max; $j++) {
      my $xc  = $i*$bin;
      my $yc  = $j*$bin;
      my $xlc = $xc+$bin;
      my $ylc = $yc+$bin;
      my $d = $density[$i][$j] ? $density[$i][$j] : 1;
      $dmax = $d > $dmax ? $d : $dmax;
      my $color_index = int(500-log($d)*$scale);
      $color_index = 5 if $color_index < 1;
      $color_index = -1 if $d < 3;
      my $color = $palette->[$color_index];
      # from percent to RGB - gna
      my $rgb = Imager::Color->new( $color->[0] * 2.55, $color->[1]*2.55, $color->[2]*2.55 );
      $image->box(color=> $rgb, xmin=>$xc,  ymin=>$yc,
                               xmax=>$xlc, ymax=>$ylc, filled=>1);

#      $image->Draw(fill=>"rgb($rgb)" , primitive=>'rectangle', points=>"$xc,$yc $xlc,$ylc");
      printf "[%s][%s] %s %s\n", $i, $j, $d, $color_index if $self->debug > 0;
    }
  }
  
  if ($self->logfile) {
    open (LOG, ">>". $self->logfile);
    printf LOG  "densitylog:  x y z pointcount: %s %s %s %s\n", $x, $y, $z, $dmax;
    close LOG;
  }
  
  $image->write(data => \$ubblob, type => 'png');
  $self->cache->set($mca, $ubblob);
  return $ubblob;
}


sub max_points_per_tile_bin {
  my ($self, $x, $y, $z) = @_;

  my $value = &Google_Tile_Factors($z, 0) ;

  my %r = Google_Tile_Calc($value, $y, $x);

  my @density;
  my $ps = $self->return_points->(\%r);
  my $max = 0;
  my $bin = $self->bin;

  foreach my $p (@$ps) {
    my @d = Google_Coord_to_Pix($value, $p->[0], $p->[1]);
    my $ix = $d[1] - $r{PXW};
    my $iy = $d[0] - $r{PYN};
    $density[int($ix/$bin)][int($iy/$bin)] ++;
    $max = $density[int($ix/$bin)][int($iy/$bin)] if $density[int($ix/$bin)][int($iy/$bin)] > $max;
  }
  return $max;
}



1;


__END__

=pod

=head1 NAME

Geo::Heatmap - generate a density map (aka heatmap) overlay layer for Google Maps 

see the www directory in the distro how it works

see the script directory for creating a scale

for a real life example see 

L<http://www.trust-box.at/dev/gm/GoogleMapsHeatmap/www/GoogleMapsHeatmap.html>

for Dokumentation see

=head1 HOMEPAGE

L<http://www.trust-box.at/googlemaps-geoheatmap/>

=head1 REQUIRES

  Moose
  CHI
  Imager

=head1 METHODS

=head2 tile

  tile();

  return the tile image in png format

=head1 ATTRIBUTES

  debug
  cache
  logfile
  return_points
  zoom_scale
  palette

=head1 USAGE

Create a Heatmap layer for GoogleMaps

=head2 The HTML part

=begin html

<pre>
<code>
  &lt;head&gt;
     &lt;meta name="viewport" content="initial-scale=1.0, user-scalable=no" /&gt;
     &lt;style type="text/css"&gt;
       html { height: 100% }
       body { height: 100%; margin: 0; padding: 0 }
       #map-canvas { height: 100% }
     &lt;/style&gt;
     &lt;script type="text/javascript"
       src="https://maps.googleapis.com/maps/api/js?key=<yourapikey>&sensor=true"&gt;
     &lt;/script&gt;
     &lt;script type="text/javascript"&gt;
       var overlayMaps = [{
         getTileUrl: function(coord, zoom) {
           return "hm.fcgi?tile="+coord.x+"+"+coord.y+"+"+zoom;
         },
 
         tileSize: new google.maps.Size(256, 256),
         isPng: true,
         opacity: 0.4
       }];
 
       function initialize() {
         var mapOptions = {
           center: new google.maps.LatLng(48.2130, 16.375),
           zoom: 9
         };
         var map = new google.maps.Map(document.getElementById("map-canvas"),
             mapOptions);
 
       var overlayMap = new google.maps.ImageMapType(overlayMaps[0]);
       map.overlayMapTypes.setAt(0,overlayMap);
 
       }
       google.maps.event.addDomListener(window, 'load', initialize);
 
     &lt;/script&gt;
   &lt;/head&gt;
   &lt;body&gt;
     &lt;div id="map-canvas"/&gt;
  &lt;/body&gt;
</code>
</pre>
<br>

=end html

=head2 The (f)cgi part

=begin html

<pre>
<code>
  #!/usr/bin/env perl
  
  use strict;
  use FCGI;
  use DBI;
  use CHI;
  use FindBin qw/$Bin/;
  use lib "$Bin/../lib";
  
  use Geo::Heatmap;
  
  #my $cache = CHI->new( driver  => 'Memcached::libmemcached',
  #    servers    => [ "127.0.0.1:11211" ],
  #    namespace  => 'GoogleMapsHeatmap',
  #);
  
  
  my $cache = CHI->new( driver => 'File',
           root_dir => '/tmp/GoogleMapsHeatmap'
       );
  
  
  our $dbh = DBI->connect("dbi:Pg:dbname=gisdb", 'gisdb', 'gisdb', {AutoCommit => 0});
  
  my $request = FCGI::Request();
  
  while ($request->Accept() >= 0) {
    my $env = $request->GetEnvironment();
    my $p = $env->{'QUERY_STRING'};
  
    my ($tile) = ($p =~ /tile=(.+)/);
    $tile =~ s/\+/ /g;
  
    # package needs a CHI Object for caching
    #               a Function Reference to get LatLOng within a Google Tile
    #               maximum number of points per zoom level
  
    my $ghm = Geo::Heatmap->new();
    $ghm->palette('palette.store');
    $ghm->zoom_scale( {
      1 => 298983,
      2 => 177127,
      3 => 104949,
      4 => 90185,
      5 => 70338,
      6 => 37742,
      7 => 28157,
      8 => 12541,
      9 => 3662,
      10 => 1275,
      11 => 417,
      12 => 130,
      13 => 41,
      14 => 18,
      15 => 10,
      16 => 6,
      17 => 2,
      18 => 0,
    } );
  
  sub get_points {
    my $r = shift;
  
    my $sth = $dbh->prepare( qq(select ST_AsEWKT(geom) from geodata
                           where geom &&
                ST_SetSRID(ST_MakeBox2D(ST_Point($r->{LATN}, $r->{LNGW}),
                                        ST_Point($r->{LATS}, $r->{LNGE})
                          ),4326))
                );
  
    $sth->execute();
  
    my @p;
    while (my @r = $sth->fetchrow) {
      my ($x, $y) = ($r[0] =~/POINT\((.+?) (.+?)\)/);
      push (@p, [$x ,$y]);
    }
    $sth->finish;
    return \@p;
  }
  </code>
  </pre>

=end html 

You need a color palette (one is included) to encode values to colors, in Storable Format as 
an arrayref of arrayrefs eg

    [50] = [34, 45, 56]

which means that a normalized value of 50 would lead to an RGB color of 34% red , 45% blue, 
56% green.

=over 4

=item zoom_scale

The maximum number of points for a given google zoom scale. You would be able to extract 
to values from the denisity log or derive them from your data in some cunning way

=item cache

You need some caching for the tiles otherwise the map would be quite slow. Use a CHI object 
with the cache you like

=item return_points

A function reference which expects a single hashref as a parameter which defines two LAT/LONG 
points to get all data points within this box:

      $r->{LATN}, $r->{LNGW}), $r->{LATS}, $r->{LNGE}

The function has to return an arrayref of arrayrefs of the points within the box

=item tile

Returns the rendered image

=back


=head1 REPOSITORY

L<https://github.com/MarkHofstetter/GoogleMapsHeatmap>

=head1 AUTHOR

Mark Hofstetter <hofstettm@cpan.org>

  Thanks to 
  brian d foy
  Marcel Gruenauer
  David Steinbrunner

=head1 TODO

=begin html

<ul>
  <li> put more magic in calculation of zoom scales </li>
  <li> make more things configurable </li>
  <li> add even more tests </li>
</ul>

=end html

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Mark Hofstetter

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

